go/store/nbs/journal_writer.go

```diff
@@ -142,14 +149,72 @@ var _ io.Closer = &journalWriter{}
 func (wr *journalWriter) bootstrapJournal(ctx context.Context) (last hash.Hash, err error) {
 	wr.lock.Lock()
 	defer wr.lock.Unlock()
-	wr.off, err = processJournalRecords(ctx, wr.file, func(o int64, r journalRec) error {
+
+	if wr.maxNovel == 0 {
+		wr.maxNovel = journalIndexDefaultMaxNovel
+	}
+	wr.ranges = newRangeIndex()
+
+	p := filepath.Join(filepath.Dir(wr.path), journalIndexFileName)
+	var ok bool
+	ok, err = fileExists(p)
+	if err != nil {
+		return
+	} else if ok {
+		wr.index, err = os.OpenFile(p, os.O_RDWR, 0666)
+	} else {
+		wr.index, err = os.OpenFile(p, os.O_RDWR|os.O_CREATE, 0666)
+	}
+	if err != nil {
+		return
+	}
+
+	if ok {
+		var info os.FileInfo
+		if info, err = wr.index.Stat(); err != nil {
+			return hash.Hash{}, err
+		}
+		err = processIndexRecords(ctx, wr.index, info.Size(), func(o int64, r indexRec) (err error) {
+			switch r.kind {
+			case tableIndexRecKind:
+				// |r.stop| is expected to point to a root hash record in |wr.journal|
+				// containing a hash equal to |r.lastRoot|, validate this here
+				var h hash.Hash
+				if h, err = peekRootHashAt(wr.journal, int64(r.end)); err != nil {
+					return err
+				} else if h != r.lastRoot {
+					return fmt.Errorf("invalid index record hash (%s != %s)", h.String(), r.lastRoot.String())
+				}
+				// populate range hashmap
+				for _, l := range deserializeLookups(r.payload) {
+					wr.ranges.put(l.a, l.r)
+				}
+				// record a high-water-mark for the indexed portion of the journal
+				wr.indexed = int64(r.end)
+				// todo: uncompressed size
+			default:
+				return fmt.Errorf("unknown index record kind (%d)", r.kind)
```

Comment:
reading an unknown record kind will throw an error, causing the index to be deleted in `corruptIndexRecovery`... probably not the behavior we want